Plane: tailsitter: always give full priority to pitch in mixing, calculate min throttle based on outflow velocity

IamPete1 opened the pull request:
A couple of improvements for tailsitters.

Firstly taisitters now do there own elevon and V tail mixing, this allows them to give full priority to pitch, any remaining range is given to yaw/roll. Unlike the existing mixer the reaming range is used symmetrically so yaw does not cancel out pitch. 
Fixes https://github.com/ArduPilot/ardupilot/issues/9412

A min outflow speed parameter has been added for use in conjunction with the disk loading gain scaling option and parameter. This will give a varying minimum throttle level depending on airspeed. This prevents difficulty when descending, but still allows low throttle at other times. 
Fixes https://github.com/ArduPilot/ardupilot/issues/16322

Finally a change to makes sure elevator/auleron/rudder are configured before setting motor limit flags based on there value.

tridge commented:
this looks good, thanks! Main thing I'd like to see is some information on testing on non-tailsitters, as this impacts the code for elevon plains and differential spoilers on non-tailsitters

IamPete1 commented on ArduPlane/tailsitter.cpp:
```diff
@@ -399,9 +409,34 @@ void Tailsitter::output(void)
+    } else {
+        aileron_mix = 0.0;
+        rudder_mix = 0.0;
+        pitch_lim = true;
```
```suggestion
        roll_lim = true;
        pitch_lim = true;
        yaw_lim = true;

```

IamPete1 commented on ArduPlane/tailsitter.cpp:
```diff
@@ -634,6 +670,25 @@ void Tailsitter::speed_scaling(void)
+                    float reverse_airspeed = 0.0;
+                    Vector3f vel;
+                    if (quadplane.ahrs.get_velocity_NED(vel)) {
+                        reverse_airspeed = quadplane.ahrs.earth_to_body(vel - quadplane.ahrs.wind_estimate()).x;
```
I wonder if we might see some benefit using this same estimate for revere airspeed for control surface scaling rather than just throttle level. 

Its a tricky one, since were calculating a squared value, we cant assume a positive answer. We would end up with control surface inversion at some point. Of course that is physically the correct thing to do, however would end up with a threshold that moves all the servos from min one way and may the other. Oscillating about that point will be nasty. It is unlikely the PIDs will be tuned right for flying backwards in anycase. 

Hopefully with this min outflow speed we should never end up with negative airflow on the control surfaces in anycase.

IamPete1 commented on ArduPlane/tailsitter.cpp:
```diff
@@ -634,6 +672,25 @@ void Tailsitter::speed_scaling(void)
+                    disk_loading_min_throttle = (((sq(disk_loading_min_outflow) + sq(reverse_airspeed)) * (0.5 * rho)) / (disk_loading.get() * GRAVITY_MSS)) * hover_throttle;
+                }
+                // AP_MotorsTailsitter has a pointer to this value
+                disk_loading_min_throttle = MAX(disk_loading_min_throttle, 0.0);
```
This would stop it being set so high that the vehicle fly's off. 
```suggestion
                disk_loading_min_throttle = MIN(MAX(disk_loading_min_throttle, 0.0), hover_throttle);
```

IamPete1 replied:
Although the throttle mix limit does a similar thing in motors, so we probably don't need it here too.

IamPete1 commented:
rebased and moved to a set method on tailistter motors rather than the slightly odd pointer to float method I used originally. This is easier to understand. In order to avoid having to add a virtual function to all motors classes I have used a direct pointer to tailsitter motors, this saves about 100 bytes or so.

IamPete1 commented on ArduPlane/tailsitter.h:
```diff
@@ -65,6 +66,10 @@ friend class Plane;
 
     // true when flying a tilt-vectored tailsitter
     bool _is_vectored;
+    // true is outputs are configured
```
these should be private

IamPete1 commented:
I'm still a little on the fence about the min outflow throttle limit stuff. It does not fix the issue completely, but it works better than anything else I have tried.
